sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -121,6 +131,77 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         if canImplicitlyCast(fromExp, toType, literalType) =>
       simplifyNumericComparison(be, fromExp, toType, value)
 
+    // As the analyzer makes sure that the list of In is already of the same data type, then the
+    // rule can simply check the first literal in `in.list` can implicitly cast to `toType` or not,
+    // and note that:
```

Comment:
The null literal handling is a bit tricky and I think it's better to put all the steps in the comment. There are 3 kinds of literals in the list:
1. null literals
2. The literals that can cast to `fromExp.dataType`
3. The literals that cannot cast to `fromExp.dataType`

null literals is special, because if we call `unwrapCast` directly, null means cannot cast, which is misleading as we can cast null literals to any data type.

The ideal steps in my mind:
1. Call `unwrapCast` with non-null literals in the list
2. If there is no literal that can cast to `fromExp.dataType`, return the original expression
3. Otherwise, use the literals that can cast to `fromExp.dataType` and the null literals to create a new `In` expression, with additional `falseIfNotNull` check.

Reply:
Thanks for your suggestion, I'll try your idea.

Reply:
Another idea: We can use `AnsiCast(...).eval`, which fails if overflow happens. Then null literal is the same as other literals that can cast to `fromExp.dataType`, and we don't need to distinguish them.

Reply:
> Another idea: We can use AnsiCast(...).eval, which fails if overflow happens. Then null literal is the same as other literals that can cast to fromExp.dataType, and we don't need to distinguish them.

Good idea,

>The ideal steps in my mind:
> 1. Call unwrapCast with non-null literals in the list
> 2. If there is no literal that can cast to fromExp.dataType, return the original expression
> 3. Otherwise, use the literals that can cast to fromExp.dataType and the null literals to create a new In expression, with additional falseIfNotNull check.

but I think this one is more clear.

Another question for stage 2: we should not return original expression when the can not cast literals is non-empty, but return `falseIfNotNull` instead, this is because original expression `In(Cast(fromExp, toType), list)` can't be optimized by other rules and trigger a filter stage, but `falseIfNotNull` can be optimized by `BooleanSimplification`, WDYT?

Reply:
SGTM, it's still useful to optimize to `falseIfNotNull`